File: Examples/intel_examples/examples_core_c/vmlc/source/vzacos.c
/*******************************************************************************
* Copyright 2001-2020 Intel Corporation.
*
* This software and the related documents are Intel copyrighted  materials,  and
* your use of  them is  governed by the  express license  under which  they were
* provided to you (License).  Unless the License provides otherwise, you may not
* use, modify, copy, publish, distribute,  disclose or transmit this software or
* the related documents without Intel's prior written permission.
*
* This software and the related documents  are provided as  is,  with no express
* or implied  warranties,  other  than those  that are  expressly stated  in the
* License.
*******************************************************************************/

/*
!  Content:
!    vzAcos  Example Program Text
!******************************************************************************/

#include <stdio.h>
#include "mkl_vml.h"

#include "_rms.h"

#define INCA 3
#define INCB 5

int main()
{
  MKL_Complex16 zA[10],zB[10];
  MKL_Complex16 zBha0[10],zBha1[10],zBha2[10];
  MKL_Complex16           zBla1[10],zBla2[10];
  MKL_Complex16           zBep1[10],zBep2[10];
  float CurRMS,MaxRMS=0.0;

  MKL_Complex16 zA_I[10*INCA],zB_I[10*INCB];
  MKL_Complex16 zBha0_I[10*INCB],zBha1_I[10*INCB],zBha2_I[10*INCB];
  MKL_Complex16                  zBla1_I[10*INCB],zBla2_I[10*INCB];
  MKL_Complex16                  zBep1_I[10*INCB],zBep2_I[10*INCB];
  float CurRMS_I,MaxRMS_I=0.0;
  MKL_INT inca=INCA,incb=INCB;

  MKL_INT i=0,vec_len=10;

  zA[0].real=-0.9000;zA[0].imag=0.9000;
  zA[1].real=-0.7000;zA[1].imag=0.7000;
  zA[2].real=-0.5000;zA[2].imag=0.5000;
  zA[3].real=-0.2999;zA[3].imag=0.2999;
  zA[4].real=-0.0999;zA[4].imag=0.0999;
  zA[5].real=0.0999;zA[5].imag=-0.0999;
  zA[6].real=0.3000;zA[6].imag=-0.3000;
  zA[7].real=0.5000;zA[7].imag=-0.5000;
  zA[8].real=0.7000;zA[8].imag=-0.7000;
  zA[9].real=0.9000;zA[9].imag=-0.9000;
  zB[0].real=2.2123245753427025e+000;zB[0].imag=-9.6595636090332149e-001;
  zB[1].real=2.1393931242264199e+000;zB[1].imag=-7.5646598100109175e-001;
  zB[2].real=2.0230747739460875e+000;zB[2].imag=-5.3063753095251787e-001;
  zB[3].real=1.8610635913216587e+000;zB[3].imag=-3.0809589760000972e-001;
  zB[4].real=1.6703610433715261e+000;zB[4].imag=-1.0022931432671847e-001;
  zB[5].real=1.4712316102182672e+000;zB[5].imag=1.0022931432671847e-001;
  zB[6].real=1.2804390679875004e+000;zB[6].imag=3.0820353078157842e-001;
  zB[7].real=1.1185178796437059e+000;zB[7].imag=5.3063753095251787e-001;
  zB[8].real=1.0021995293633732e+000;zB[8].imag=7.5646598100109175e-001;
  zB[9].real=9.2926807824709057e-001;zB[9].imag=9.6595636090332149e-001;

  for(i=0;i<10;i++) {
    zA_I[i*inca]=zA[i];
    zB_I[i*incb]=zB[i];
  }

  vzAcos(vec_len,zA,zBha0);
  vzAcosI(vec_len,zA_I,inca,zBha0_I,incb);

  vmzAcos(vec_len,zA,zBep1,VML_EP);
  vmzAcosI(vec_len,zA_I,inca,zBep1_I,incb,VML_EP);

  vmlSetMode(VML_EP);
  vzAcos(vec_len,zA,zBep2);
  vzAcosI(vec_len,zA_I,inca,zBep2_I,incb);

  vmzAcos(vec_len,zA,zBla1,VML_LA);
  vmzAcosI(vec_len,zA_I,inca,zBla1_I,incb,VML_LA);

  vmlSetMode(VML_LA);
  vzAcos(vec_len,zA,zBla2);
  vzAcosI(vec_len,zA_I,inca,zBla2_I,incb);

  vmzAcos(vec_len,zA,zBha1,VML_HA);
  vmzAcosI(vec_len,zA_I,inca,zBha1_I,incb,VML_HA);

  vmlSetMode(VML_HA);
  vzAcos(vec_len,zA,zBha2);
  vzAcosI(vec_len,zA_I,inca,zBha2_I,incb);

  for(i=0;i<10;i++) {
    if(zBha0[i].real!=zBha1[i].real || zBha0[i].imag!=zBha1[i].imag || zBha1[i].real!=zBha2[i].real ||
        zBha1[i].imag!=zBha2[i].imag) {
      printf("Error! Difference between vzAcos and vmzAcos in VML_HA mode detected.\n");
      return 1;
    }

    if(zBha0_I[i*incb].real!=zBha1_I[i*incb].real || zBha0_I[i*incb].imag!=zBha1_I[i*incb].imag ||
        zBha1_I[i*incb].real!=zBha2_I[i*incb].real || zBha1_I[i*incb].imag!=zBha2_I[i*incb].imag) {
      printf("Error! Difference between vzAcosI and vmzAcosI in VML_HA mode detected.\n");
      return 1;
    }

    if(zBla1[i].real!=zBla2[i].real || zBla1[i].imag!=zBla2[i].imag) {
      printf("Error! Difference between vzAcos and vmzAcos in VML_LA mode detected.\n");
      return 1;
    }

    if(zBla1_I[i*incb].real!=zBla2_I[i*incb].real || zBla1_I[i*incb].imag!=zBla2_I[i*incb].imag) {
      printf("Error! Difference between vzAcosI and vmzAcosI in VML_LA mode detected.\n");
      return 1;
    }

    if(zBep1[i].real!=zBep2[i].real || zBep1[i].imag!=zBep2[i].imag) {
      printf("Error! Difference between vzAcos and vmzAcos in VML_EP mode detected.\n");
      return 1;
    }

    if(zBep1_I[i*incb].real!=zBep2_I[i*incb].real || zBep1_I[i*incb].imag!=zBep2_I[i*incb].imag) {
      printf("Error! Difference between vzAcosI and vmzAcosI in VML_EP mode detected.\n");
      return 1;
    }
  }

  printf("vzAcos test/example program\n\n");
  printf("           Argument                           vzAcos\n");
  printf("===============================================================================\n");
  for(i=0;i<10;i++) {
    printf("   % .4f %+.4f*i      % .10f %+.10f*i\n",zA[i].real,zA[i].imag,zBha0[i].real,zBha0[i].imag);
    CurRMS=zrelerr(zB[i],zBha0[i]);
    if(CurRMS>MaxRMS) MaxRMS=CurRMS;
  }
  printf("\n");
  printf("vzAcosI test/example program\n\n");
  printf("           Argument                           vzAcos\n");
  printf("===============================================================================\n");
  for(i=0;i<10;i++) {
    printf("   % .4f %+.4f*i      % .10f %+.10f*i\n",zA_I[i*inca].real,zA_I[i*inca].imag,zBha0_I[i*incb].real,
        zBha0_I[i*incb].imag);
    CurRMS_I=zrelerr(zB_I[i*incb],zBha0_I[i*incb]);
    if(CurRMS_I>MaxRMS_I) MaxRMS_I=CurRMS_I;
  }
  printf("\n");
  if(MaxRMS>=1e-14) {
    printf("Error! Relative accuracy is %.16f\n",MaxRMS);
    return 1;
  }
  else {
    printf("Relative accuracy is %.16f\n",MaxRMS);
  }

  printf("\n");

  if(MaxRMS_I>=1e-14) {
    printf("Error! Relative strided accuracy is %.16f\n",MaxRMS_I);
    return 1;
  }
  else {
    printf("Relative strided accuracy is %.16f\n",MaxRMS_I);
  }
  return 0;
}

File: Examples/intel_examples/examples_core_f/vmlf/source/vzacos.f
!===============================================================================
! Copyright 2001-2020 Intel Corporation.
!
! This software and the related documents are Intel copyrighted  materials,  and
! your use of  them is  governed by the  express license  under which  they were
! provided to you (License).  Unless the License provides otherwise, you may not
! use, modify, copy, publish, distribute,  disclose or transmit this software or
! the related documents without Intel's prior written permission.
!
! This software and the related documents  are provided as  is,  with no express
! or implied  warranties,  other  than those  that are  expressly stated  in the
! License.
!===============================================================================

!  Content:
!    vzAcos  Example Program Text
!*******************************************************************************

      include "_rms.fi"

      program MKL_VML_TEST

      include "mkl_vml.f90"
      real(kind=4) :: srelerr
      real(kind=8) :: drelerr
      real(kind=4) :: crelerr
      real(kind=8) :: zrelerr

      complex(kind=8) zA(10)
      complex(kind=8) zB(10)
      complex(kind=8) zBha0(10)
      complex(kind=8) zBha1(10)
      complex(kind=8) zBha2(10)
      complex(kind=8) zBla1(10)
      complex(kind=8) zBla2(10)
      complex(kind=8) zBep1(10)
      complex(kind=8) zBep2(10)

      real(kind=8) CurRMS,MaxRMS

      integer, parameter :: inca=3
      integer, parameter :: incb=5
      complex(kind=8) zA_I(10*inca)
      complex(kind=8) zB_I(10*incb)
      complex(kind=8) zBha0_I(10*incb)
      complex(kind=8) zBha1_I(10*incb)
      complex(kind=8) zBha2_I(10*incb)
      complex(kind=8) zBla1_I(10*incb)
      complex(kind=8) zBla2_I(10*incb)
      complex(kind=8) zBep1_I(10*incb)
      complex(kind=8) zBep2_I(10*incb)
      real(kind=8) CurRMS_I,MaxRMS_I

      integer(kind=8) mode
      integer tmode
      integer i, vec_len

      vec_len=10
      MaxRMS=0.0
      MaxRMS_I=0.0

      zA( 1)=(-0.9000d+000,0.9000d+000)
      zA( 2)=(-0.7000d+000,0.7000d+000)
      zA( 3)=(-0.5000d+000,0.5000d+000)
      zA( 4)=(-0.2999d+000,0.2999d+000)
      zA( 5)=(-0.0999d+000,0.0999d+000)
      zA( 6)=(0.0999d+000,-0.0999d+000)
      zA( 7)=(0.3000d+000,-0.3000d+000)
      zA( 8)=(0.5000d+000,-0.5000d+000)
      zA( 9)=(0.7000d+000,-0.7000d+000)
      zA(10)=(0.9000d+000,-0.9000d+000)
      zB( 1)=(2.2123245753427025d+000,-9.6595636090332149d-001)
      zB( 2)=(2.1393931242264199d+000,-7.5646598100109175d-001)
      zB( 3)=(2.0230747739460875d+000,-5.3063753095251787d-001)
      zB( 4)=(1.8610635913216587d+000,-3.0809589760000972d-001)
      zB( 5)=(1.6703610433715261d+000,-1.0022931432671847d-001)
      zB( 6)=(1.4712316102182672d+000,1.0022931432671847d-001)
      zB( 7)=(1.2804390679875004d+000,3.0820353078157842d-001)
      zB( 8)=(1.1185178796437059d+000,5.3063753095251787d-001)
      zB( 9)=(1.0021995293633732d+000,7.5646598100109175d-001)
      zB(10)=(9.2926807824709057d-001,9.6595636090332149d-001)

      do i=1,10
          zA_I((i-1)*inca+1)=zA(i)
          zB_I((i-1)*incb+1)=zB(i)
      end do

      call VZACOS(vec_len,zA,zBha0)
      call VZACOSI(vec_len,zA_I,inca,zBha0_I,incb)

      mode=VML_EP
      call VMZACOS(vec_len,zA,zBep1,mode)
      call VMZACOSI(vec_len,zA_I,inca,zBep1_I,incb,mode)

      tmode=VML_EP
      tmode=VMLSETMODE(tmode)
      call VZACOS(vec_len,zA,zBep2)
      call VZACOSI(vec_len,zA_I,inca,zBep2_I,incb)

      mode=VML_LA
      call VMZACOS(vec_len,zA,zBla1,mode)
      call VMZACOSI(vec_len,zA_I,inca,zBla1_I,incb,mode)

      tmode=VML_LA
      tmode=VMLSETMODE(tmode)
      call VZACOS(vec_len,zA,zBla2)
      call VZACOSI(vec_len,zA_I,inca,zBla2_I,incb)

      mode=VML_HA
      call VMZACOS(vec_len,zA,zBha1,mode)
      call VMZACOSI(vec_len,zA_I,inca,zBha1_I,incb,mode)

      tmode=VML_HA
      tmode=VMLSETMODE(tmode)
      call VZACOS(vec_len,zA,zBha2)
      call VZACOSI(vec_len,zA_I,inca,zBha2_I,incb)

      do i=1,10
          if(zBha0(i) .ne. zBha1(i)) then
              print *,"Error! Difference between VZACOS and"
              print *," VMZACOS in VML_HA mode detected"
              stop 1
          endif
          if(zBha0_I((i-1)*incb+1) .ne. zBha1_I((i-1)*incb+1)) then
              print *,"Error! Difference between VZACOSI and"
              print *," VMZACOSI in VML_HA mode detected"
              stop 1
          endif
          if(zBha1(i) .ne. zBha2(i)) then
              print *,"Error! Difference between VZACOS and"
              print *," VMZACOS in VML_HA mode detected"
              stop 1
          endif
          if(zBha1_I((i-1)*incb+1) .ne. zBha2_I((i-1)*incb+1)) then
              print *,"Error! Difference between VZACOSI and"
              print *," VMZACOSI in VML_HA mode detected"
              stop 1
          endif
          if(zBla1(i) .ne. zBla2(i)) then
              print *,"Error! Difference between VZACOS and"
              print *," VMZACOS in VML_LA mode detected"
              stop 1
          endif
          if(zBla1_I((i-1)*incb+1) .ne. zBla2_I((i-1)*incb+1)) then
              print *,"Error! Difference between VZACOSI and"
              print *," VMZACOSI in VML_LA mode detected"
              stop 1
          endif
          if(zBep1(i) .ne. zBep2(i)) then
              print *,"Error! Difference between VZACOS and"
              print *," VMZACOS in VML_EP mode detected"
              stop 1
          endif
          if(zBep1_I((i-1)*incb+1) .ne. zBep2_I((i-1)*incb+1)) then
              print *,"Error! Difference between VZACOSI and"
              print *," VMZACOSI in VML_EP mode detected"
              stop 1
          endif
      end do

      print *,"vzAcos test/example program"
      print *,""
      print *,"           Argument                     vzAcos"
      print *,"======================================================", &
     &        "========================"
      do i=1,vec_len
            print 10,zA(i),"    ",zBha0(i)
            CurRMS=zrelerr(zB(i),zBha0(i))
            if(CurRMS>MaxRMS) MaxRMS=CurRMS
      end do

      print *,""
      print *,"vzAcosI test/example program"
      print *,"           Argument                     vzAcos"
      print *,"======================================================", &
     &        "========================"
      do i=1,10
            print 10,zA_I((i-1)*inca+1),"    ",zBha0_I((i-1)*incb+1)
            CurRMS_I=zrelerr(zB_I((i-1)*incb+1),zBha0_I((i-1)*incb+1))
            if(CurRMS_I>MaxRMS_I) MaxRMS_I=CurRMS_I
      end do

      print *,""
      if(MaxRMS>=1e-14) then
            print 11,"Error! Relative accuracy is ",MaxRMS
            stop 1
      else
            print 11,"Relative accuracy is ",MaxRMS
      endif

      if(MaxRMS_I>=1e-14) then
            print 11,"Error! Relative strided accuracy is ",MaxRMS_I
            stop 1
      else
            print 11,"Relative strided accuracy is ",MaxRMS_I
      endif

10    format(E15.7,E15.7,A5,E15.7,E15.7)
11    format(A,F25.16)

      end
